packages/driver-azure/src/driver/index.ts

```diff
@@ -0,0 +1,272 @@
+import { InferredFlags } from '@oclif/core/lib/interfaces'
+import { Resource, VirtualMachine } from '@azure/arm-compute'
+import { inspect } from 'util'
+import { DefaultAzureCredential } from '@azure/identity'
+import { SubscriptionClient } from '@azure/arm-subscriptions'
+import {
+  generateSshKeyPair,
+  Machine,
+  MachineCreationDriver,
+  MachineCreationDriverFactory,
+  MachineDriver,
+  MachineDriverFactory,
+  PartialMachine,
+  telemetryEmitter,
+} from '@preevy/core'
+import { client, REGIONS } from './client'
+import { CUSTOMIZE_BARE_MACHINE } from './scripts'
+import { AzureCustomTags, extractResourceGroupNameFromId } from './vm-creation-utils'
+
+type RootObjectDetailsError = {
+  code: string
+  message: string
+}
+type RootObjectDetails = {
+  error: RootObjectDetailsError
+}
+
+export type AzureErrorResponse = {
+  details: RootObjectDetails
+  name: string
+  code: string
+  statusCode: number
+  message: string
+}
+
+type DriverContext = {
+  profileId: string
+  region: string
+  subscriptionId: string
+}
+
+type MachineCreationContext = DriverContext & {
+  vmSize?: string
+  resourceGroupId: string
+}
+
+const UBUNTU_IMAGE_DETAILS = {
+  publisher: 'Canonical',
+  offer: '0001-com-ubuntu-server-jammy',
+  sku: '22_04-lts-gen2',
+}
+
+const requireTagValue = (tags: Resource['tags'], key: string) => {
+  if (!tags?.[key]) {
+    throw new Error(`Could not find required tag ${key} in ${inspect(tags)}`)
+  }
+  return tags[key]
+}
+
+const machineFromInstance = (
+  { privateIPAddress, publicIPAddress, vm }: {
+    privateIPAddress: string
+    publicIPAddress: string
+    vm: VirtualMachine}
+): Machine & { envId: string } => {
+  if (!vm.id || !vm.osProfile?.adminUsername) {
+    throw new Error('Could not create a machine from instance')
+  }
+  return {
+    privateIPAddress,
+    publicIPAddress,
+    providerId: extractResourceGroupNameFromId(vm.id),
+    sshKeyName: 'default',
+    sshUsername: vm.osProfile.adminUsername,
+    version: '',
+    envId: requireTagValue(vm.tags, AzureCustomTags.ENV_ID),
+  }
+}
+
+const machineDriver = ({ region, subscriptionId, profileId }: DriverContext): MachineDriver => {
+  const cl = client({
+    region,
+    subscriptionId,
+    profileId,
+  })
+
+  return {
+    customizationScripts: CUSTOMIZE_BARE_MACHINE,
+    friendlyName: 'Microsoft Azure',
+    getMachine: async ({ envId }) => cl.getInstance(envId).then(vm => machineFromInstance(vm)),
+    listMachines: async () => {
+      const vmInstances = await asyncToArray(cl.listInstances())
+      const allResourceGroups = cl.listResourceGroups()
+      return asyncMap(async resourceGroup => {
+        const vm = typeof resourceGroup.name === 'string' && vmInstances.find(i => extractResourceGroupNameFromId(i.vm.id as string) === (resourceGroup.name as string).toLowerCase())
+        if (vm) {
+          return { ...machineFromInstance(vm),
+            partial: false } as (Machine|PartialMachine) & { envId: string }
```

Comment:
why is `partial` needed?

Reply:
it's a list that contains a machine or partial machine.
it's being filtered by type more easily.
what do you suggest? a machine with optional props?

Reply:
correct me if I'm wrong - deducing whether it's partial is just a matter of checking for other props, so the `partial` prop doesn't add new information and can just cause mismatching bugs.